File: src/core/stdlib/series/extremes.ts
/**
 * The lookback extremes: `highest`, `lowest` and the two that report where the
 * extreme was set.
 *
 * `highestBars` and `lowestBars` count bars back, 0 being this bar, which is
 * what `stdlib.md` section 9 says they return. A tie goes to the most recent
 * bar: the lookback's high was set most recently at that bar, and reporting the
 * older one would make the answer jump backwards as an equal high rolls in.
 */
import type { Series, StateRecord, Tail, Value } from '../values/index.js';
import { NONE, fold, result, ring, tailOf } from '../values/index.js';

/** Where the extreme sits in a complete lookback, as bars back from this one. */
function extremeBack(
  read: (back: number) => Value,
  len: number,
  wantHigh: boolean,
): number {
  let best = read(len - 1) as number;
  let bestBack = len - 1;
  // Oldest first, so an equal value later in the lookback replaces the earlier
  // one and the answer is the most recent bar that set the extreme.
  for (let position = 1; position < len; position += 1) {
    const back = len - 1 - position;
    const value = read(back) as number;
    if (wantHigh ? value >= best : value <= best) {
      best = value;
      bestBack = back;
    }
  }
  return bestBack;
}

/** `highest`, `lowest` and the two that report where the extreme was set. */
export function extremeStep(
  state: StateRecord,
  key: string,
  value: Value,
  len: number | null,
  wantHigh: boolean,
  asBars: boolean,
): Value {
  const lookback = ring(state, key, len);
  lookback.push(value);
  if (len === null || !lookback.complete()) return NONE;
  const back = extremeBack((k) => lookback.at(k), len, wantHigh);
  return asBars ? back : result(lookback.at(back) as number);
}

function extremeTail(len: number, wantHigh: boolean, asBars: boolean): Tail<Value, Value> {
  return tailOf((state, value: Value) => extremeStep(state, 'q', value, len, wantHigh, asBars));
}

/** `highest(src, len)`: the largest value in the last `len` bars, from bar `len - 1`. */
export function highestTail(len: number): Tail<Value, Value> {
  return extremeTail(len, true, false);
}

/** `highest(src, len)` over a whole series. */
export function highest(src: Series, len: number): Value[] {
  return fold(highestTail(len), src);
}

/** `lowest(src, len)`: the smallest value in the last `len` bars, from bar `len - 1`. */
export function lowestTail(len: number): Tail<Value, Value> {
  return extremeTail(len, false, false);
}

/** `lowest(src, len)` over a whole series. */
export function lowest(src: Series, len: number): Value[] {
  return fold(lowestTail(len), src);
}

/** `highestBars(src, len)`: how many bars back the lookback's high was set. */
export function highestBarsTail(len: number): Tail<Value, Value> {
  return extremeTail(len, true, true);
}

/** `highestBars(src, len)` over a whole series. */
export function highestBars(src: Series, len: number): Value[] {
  return fold(highestBarsTail(len), src);
}

/** `lowestBars(src, len)`: how many bars back the lookback's low was set. */
export function lowestBarsTail(len: number): Tail<Value, Value> {
  return extremeTail(len, false, true);
}

/** `lowestBars(src, len)` over a whole series. */
export function lowestBars(src: Series, len: number): Value[] {
  return fold(lowestBarsTail(len), src);
}

File: engine/openscript/library/extremes.py
"""The window scans of `stdlib.md` sections 9 and 20.10, and the pivots.

Section 20.10 names these as the functions with no accumulation order to fix:
they select a value and compute nothing, so an implementer can write the scan the
obvious way and be bit identical. **The one thing a scan can differ on is a
tie**, and the tie rules are section 9's, so they are what this file is careful
about.

- An extreme goes to **the most recent bar that set it**, so an equal value later
  in the window replaces the earlier one. That only shows in the two calls that
  report an age, and it shows there on every flat stretch of a series.
- A pivot is **strict on both sides**, so a run of equal values holds no pivot at
  all. A test written with "at or above" would report a pivot at each end of a
  flat top, which is two pivots where the chart shows none.

A window arrives newest first, so an index into it is an age in bars and
``held[0]`` is what this bar contributed.
"""

from typing import Optional, Sequence

from .series import Region, contributed, window
from .values import ABSENT, Value, number, result


def high_at(held: Sequence[float]) -> int:
    """The age in bars of the window's high, the most recent bar that set it.

    The scan runs oldest first and replaces on an equal value, which is how the
    tie reaches the newest of the bars that hold it rather than the oldest.
    """
    at = len(held) - 1
    for age in range(len(held) - 2, -1, -1):
        if held[age] >= held[at]:
            at = age
    return at


def low_at(held: Sequence[float]) -> int:
    """The age in bars of the window's low, on the same terms."""
    at = len(held) - 1
    for age in range(len(held) - 2, -1, -1):
        if held[age] <= held[at]:
            at = age
    return at


def window_high(values: Sequence[Value], length: Optional[int]) -> Value:
    """The outright high of the last ``length`` contributions, or absence."""
    held = window(values, length)
    return ABSENT if held is None else result(held[high_at(held)])


def window_low(values: Sequence[Value], length: Optional[int]) -> Value:
    """The outright low of the last ``length`` contributions, or absence."""
    held = window(values, length)
    return ABSENT if held is None else result(held[low_at(held)])


def highest(state: Region, value: Value, length: Optional[int]) -> Value:
    """``highest(src, len)``: the largest value in the last ``len`` bars."""
    return window_high(contributed(state, "src", number(value), length), length)


def lowest(state: Region, value: Value, length: Optional[int]) -> Value:
    """``lowest(src, len)``: the smallest."""
    return window_low(contributed(state, "src", number(value), length), length)


def highest_bars(state: Region, value: Value, length: Optional[int]) -> Value:
    """``highestBars(src, len)``: how many bars back the window's high was set."""
    held = window(contributed(state, "src", number(value), length), length)
    return ABSENT if held is None else float(high_at(held))


def lowest_bars(state: Region, value: Value, length: Optional[int]) -> Value:
    """``lowestBars(src, len)``: how many bars back the window's low was set."""
    held = window(contributed(state, "src", number(value), length), length)
    return ABSENT if held is None else float(low_at(held))


def pivot_high(state: Region, value: Value, left: Optional[int], right: Optional[int]) -> Value:
    """``pivotHigh(src, left, right)``: a local high, reported ``right`` bars after it.

    Reporting it at the pivot bar would be a lookahead: the value would appear on
    history at a bar where no script could have had it. So the window is
    ``left + right + 1`` wide, the candidate sits ``right`` bars back inside it,
    and the answer is absent on every bar that is not one.
    """
    values = contributed(state, "src", number(value), _span(left, right))
    return _pivot(values, left, right, True)


def pivot_low(state: Region, value: Value, left: Optional[int], right: Optional[int]) -> Value:
    """``pivotLow(src, left, right)``: a local low, on the same terms."""
    values = contributed(state, "src", number(value), _span(left, right))
    return _pivot(values, left, right, False)


def _span(left: Optional[int], right: Optional[int]) -> Optional[int]:
    """The window a pivot test needs: the bars either side and the candidate."""
    if left is None or right is None:
        return None
    return left + right + 1


def _pivot(
    values: Sequence[Value], left: Optional[int], right: Optional[int], above: bool
) -> Value:
    """The candidate's value where it beats both sides strictly, absence otherwise."""
    held = window(values, _span(left, right))
    if held is None or right is None:
        return ABSENT
    candidate = held[right]
    for age in range(len(held)):
        if age == right:
            continue
        if above and held[age] >= candidate:
            return ABSENT
        if not above and held[age] <= candidate:
            return ABSENT
    return result(candidate)
